pkg/controller/v1alpha1/routes/route_controller.go

```diff
@@ -0,0 +1,436 @@
+			fmt.Sprintf("successfully automatically created a service `%s`", svc.Name)))
+	}
+
+	// Create Issuers
+	var issuer standardv1alpha1.TLS
+	if routeTrait.Spec.TLS == nil || routeTrait.Spec.TLS.IssuerName == "" {
+		issuerName, err := r.createSelfsignedIssuer(ctx, &routeTrait)
+		if err != nil {
+			r.record.Event(eventObj, event.Warning(errCreateIssuer, err))
+			return oamutil.ReconcileWaitResult,
+				oamutil.PatchCondition(ctx, r, &routeTrait,
+					cpv1alpha1.ReconcileError(errors.Wrap(err, errCreateIssuer)))
+		}
+		r.record.Event(eventObj, event.Normal("Issuer created",
+			fmt.Sprintf("successfully automatically created a Issuer for route TLS `%s`", issuerName)))
+		issuer.Type = standardv1alpha1.NamespaceIssuer
+		issuer.IssuerName = issuerName
+	} else {
+		issuer = *routeTrait.Spec.TLS
+	}
+
+	// Create Ingress
+	// construct the serviceMonitor that hooks the service to the prometheus server
+	ingress := constructNginxIngress(&routeTrait, issuer, svc, svcPort)
+	// server side apply the serviceMonitor, only the fields we set are touched
+	applyOpts := []client.PatchOption{client.ForceOwnership, client.FieldOwner(ingress.GetUID())}
+	if err := r.Patch(ctx, ingress, client.Apply, applyOpts...); err != nil {
+		mLog.Error(err, "Failed to apply to ingress")
+		r.record.Event(eventObj, event.Warning(errApplyNginxIngress, err))
+		return oamutil.ReconcileWaitResult,
+			oamutil.PatchCondition(ctx, r, &routeTrait,
+				cpv1alpha1.ReconcileError(errors.Wrap(err, errApplyNginxIngress)))
+	}
+	r.record.Event(eventObj, event.Normal("Nginx Ingress created",
+		fmt.Sprintf("successfully server side patched a route trait `%s`", routeTrait.Name)))
+
+	// TODO(wonderflow): GC mechanism for no used ingress, service, issuer
+
+	routeTrait.Status.Service = svc
+	routeTrait.Status.Ingress = &runtimev1alpha1.TypedReference{
+		APIVersion: v1beta1.SchemeGroupVersion.String(),
+		Kind:       reflect.TypeOf(v1beta1.Ingress{}).Name(),
+		Name:       ingress.Name,
+		UID:        routeTrait.UID,
+	}
+	return ctrl.Result{}, oamutil.PatchCondition(ctx, r, &routeTrait)
+}
+
+// create a service that targets the exposed workload pod
+func (r *Reconciler) createService(ctx context.Context, mLog logr.Logger, workload *unstructured.Unstructured,
+	routeTrait *v1alpha1.Route) (*runtimev1alpha1.TypedReference, int32, error) {
+
+	oamService := &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       common.ServiceKind,
+			APIVersion: common.ServiceAPIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "route-" + workload.GetName(),
+			Namespace: workload.GetNamespace(),
+			Labels:    oamServiceLabel,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeLoadBalancer,
+		},
+	}
+	// assign selector
+	if routeTrait.Spec.Backend != nil && routeTrait.Spec.Backend.SelectLabels != nil {
+		oamService.Spec.Selector = routeTrait.Spec.Backend.SelectLabels
+	}
+
+	port, labels, err := DiscoverPortLabel(workload)
```

Comment:
this would not work for the core containerizedWorkload or anything that doesn't have the pod.spec as is. I wonder if you want to also look at the childResources?

Reply:
yes, I added